bluemira/builders/tf_coils.py

```diff
@@ -140,17 +143,397 @@ def _make_single_circuit(self, wire):
+            Wire along which the points will be selected
+        """
+        super().set_wire(wire)
+        if self.x_frac is not None and not np.isclose(self.x_frac, 0.0):
+            self.x_frac = np.clip(self.x_frac, 0.005, 0.995)
+            bb = wire.bounding_box
+
+            x_min = bb.x_min + self.x_frac * (bb.x_max - bb.x_min)
+
+            z_min, z_max = bb.z_min - 10, bb.z_max + 10
+            cut_face = BluemiraFace(
+                make_polygon(
+                    {
+                        "x": [0, x_min, x_min, 0],
+                        "y": 0,
+                        "z": [z_min, z_min, z_max, z_max],
+                    },
+                    closed=True,
+                )
+            )
+            wire = boolean_cut(wire, cut_face)[0]
+        self.points = wire.discretize(byedges=True, ndiscr=self.n_rip_points)
+
+
+class ExtremaSelector(RipplePointSelector):
+    """
+    Select the extrema of the wire and constrain ripple there.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def set_wire(self, wire: BluemiraWire):
+        """
+        Set the wire along which the points will be selected
+
+        Parameters
+        ----------
+        wire
+            Wire along which the points will be selected
+        """
+        super().set_wire(wire)
+        coords = wire.discretize(byedges=True, ndiscr=2000)
+        arg_x_min = np.argmin(coords.x)
+        arg_x_max = np.argmax(coords.x)
+        arg_z_min = np.argmin(coords.z)
+        arg_z_max = np.argmax(coords.z)
+        self.points = Coordinates(
+            [
+                coords.points[arg_x_min],
+                coords.points[arg_x_max],
+                coords.points[arg_z_min],
+                coords.points[arg_z_max],
+            ]
+        )
+
+
+class FixedSelector(RipplePointSelector):
+    """
+    Specified points at which to constrain the ripple, overrides any information
+    relating directly to the separatrix.
+
+    Parameters
+    ----------
+    points:
+        Points at which the ripple should be constrained.
+    """
+
+    def __init__(self, points: Coordinates):
+        self.points = points
+
+
+class MaximiseSelector(RipplePointSelector):
+    """
+    Finds and constrains the maximum ripple along the specified wire during
+    each minimisation function call.
+
+    Parameters
+    ----------
+    wire:
+        Wire along which to constrain the ripple
+    """
+
+    def __init__(self):
+        self.points = None
+
+    def set_wire(self, wire):
+        """
+        Set the wire along which the points will be selected
+
+        Parameters
+        ----------
+        wire
+            Wire along which the points will be selected
+        """
+        super().set_wire(wire)
+        points = wire.discretize(byedges=True, ndiscr=200)
+        arg_x_max = np.argmax(points.x)
+        x_max_point = points[:, arg_x_max]
+        self._alpha_0 = wire.parameter_at(x_max_point, tolerance=10 * EPS)
```

Comment:
probably the same question about inconsistency with `self.wire`, should we use `self.wire` instead of `wire`?